torch/distributed/_dist2.py

```diff
@@ -13,27 +13,387 @@
 from torch._C._distributed_c10d import (
     _current_process_group,
     _set_process_group,
+    Backend,
+    AllgatherOptions,
+    AllreduceCoalescedOptions,
+    AllreduceOptions,
+    AllToAllOptions,
+    BarrierOptions,
+    BroadcastOptions,
+    GatherOptions,
     ProcessGroup,
     ReduceOp,
+    ReduceOp,
+    ReduceOptions,
+    ReduceScatterOptions,
+    ScatterOptions,
     Store,
 )
 from torch.distributed.rendezvous import rendezvous
 
 
-_BACKENDS: dict[str, "ProcessGroupFactory"] = {}
+_BACKENDS: dict[str, "CommunicatorFactory"] = {}
 
 __all__ = [
-    "ProcessGroup",
+    "Communicator",
     "ReduceOp",
-    "ProcessGroupFactory",
+    "CommunicatorFactory",
     "register_backend",
-    "new_group",
-    "current_process_group",
-    "process_group",
+    "new_comm",
+    "current_comm",
+    "comm",
 ]
 
 
-class ProcessGroupFactory(Protocol):
+class Communicator:
+    """
+    A communicator allows for communication between processes. This maps 1:1
```

Comment:
I was about to object to renaming here but I think it makes sense to draw the distinction.

That said before we adopt a new api I want to include 2 things in it
- traceable collectives
- autograd support 

We're already running into needing auto grad support for auto parallel and I want to make sure we can build it in a way that lines up with the direction of our new comm api.